File: tests/test_source_browser.py
from __future__ import annotations

import json
import socket
import threading
import unittest

from source_browser import PROTOCOL, SourceBrowser, parse_discovery_peers


class SourceBrowserTest(unittest.TestCase):
    def test_discovers_and_requests_selected_source(self) -> None:
        changed = threading.Event()
        snapshots = []

        def on_sources_changed(sources):
            snapshots.append(sources)
            changed.set()

        fake_tx = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        fake_tx.bind(("127.0.0.1", 0))
        fake_tx.settimeout(2)
        control_port = fake_tx.getsockname()[1]
        browser = SourceBrowser(
            "rx-one",
            discovery_port=0,
            stream_port=6200,
            query_targets=(),
            query_interval=60,
            on_sources_changed=on_sources_changed,
        )
        browser.start()
        try:
            announcement = {
                "protocol": PROTOCOL,
                "type": "source_announce",
                "source_id": "tx-one",
                "source_name": "Kitchen TX",
                "control_port": control_port,
                "transport": "rtp/udp",
                "encoding": "JPEG",
                "payload": 26,
            }
            fake_tx.sendto(
                json.dumps(announcement).encode(),
                ("127.0.0.1", browser.discovery_port),
            )
            self.assertTrue(changed.wait(2))
            self.assertEqual("tx-one", snapshots[-1][0].source_id)
            self.assertEqual("Kitchen TX", snapshots[-1][0].name)
            self.assertTrue(snapshots[-1][0].online)

            self.assertTrue(browser.select_source("tx-one"))
            payload, _address = fake_tx.recvfrom(4096)
            request = json.loads(payload)
            self.assertEqual("stream_request", request["type"])
            self.assertEqual("tx-one", request["source_id"])
            self.assertEqual("rx-one", request["receiver_id"])
            self.assertEqual(6200, request["stream_port"])
        finally:
            browser.stop()
            fake_tx.close()

    def test_keeps_multiple_channels_and_moves_request_when_selected(self) -> None:
        changed = threading.Event()
        fake_tx_one = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        fake_tx_two = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        fake_tx_one.bind(("127.0.0.1", 0))
        fake_tx_two.bind(("127.0.0.1", 0))
        fake_tx_one.settimeout(2)
        fake_tx_two.settimeout(2)
        browser = SourceBrowser(
            "rx-one",
            discovery_port=0,
            stream_port=6200,
            query_targets=(),
            query_interval=60,
            on_sources_changed=lambda _sources: changed.set(),
        )
        browser.start()
        try:
            for source_id, source_name, control_port in (
                ("tx-one", "Kitchen TX", fake_tx_one.getsockname()[1]),
                ("tx-two", "Workshop TX", fake_tx_two.getsockname()[1]),
            ):
                announcement = {
                    "protocol": PROTOCOL,
                    "type": "source_announce",
                    "source_id": source_id,
                    "source_name": source_name,
                    "control_port": control_port,
                    "transport": "rtp/udp",
                    "encoding": "JPEG",
                    "payload": 26,
                }
                changed.clear()
                fake_tx_one.sendto(
                    json.dumps(announcement).encode(),
                    ("127.0.0.1", browser.discovery_port),
                )
                self.assertTrue(changed.wait(2))

            self.assertEqual(
                ("tx-one", "tx-two"),
                tuple(source.source_id for source in browser.sources),
            )

            self.assertTrue(browser.select_source("tx-one"))
            request_one = json.loads(fake_tx_one.recv(4096))
            self.assertEqual("stream_request", request_one["type"])

            self.assertTrue(browser.select_source("tx-two"))
            stop_one = json.loads(fake_tx_one.recv(4096))
            request_two = json.loads(fake_tx_two.recv(4096))
            self.assertEqual("stream_stop", stop_one["type"])
            self.assertEqual("stream_request", request_two["type"])
            self.assertEqual("tx-two", browser.selected_source_id)
        finally:
            browser.stop()
            fake_tx_one.close()
            fake_tx_two.close()

    def test_parse_discovery_peers(self) -> None:
        self.assertEqual(
            (("10.0.0.10", 5601), ("tx.example", 6001)),
            parse_discovery_peers("10.0.0.10, tx.example:6001"),
        )
        with self.assertRaises(ValueError):
            parse_discovery_peers("tx.example:70000")


if __name__ == "__main__":
    unittest.main()

File: native/tests/test_source_browser.cpp
#include "gar_stream_rx/source_browser.hpp"

#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <unistd.h>

#include <cassert>
#include <cerrno>
#include <chrono>
#include <condition_variable>
#include <cstring>
#include <mutex>
#include <stdexcept>
#include <string>
#include <string_view>
#include <vector>

namespace {

int create_server(std::uint16_t& port) {
    const int descriptor = socket(AF_INET, SOCK_DGRAM | SOCK_CLOEXEC, 0);
    if (descriptor < 0) {
        throw std::runtime_error(std::strerror(errno));
    }
    timeval timeout{2, 0};
    setsockopt(descriptor, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));
    sockaddr_in address{};
    address.sin_family = AF_INET;
    address.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
    address.sin_port = 0;
    if (bind(descriptor, reinterpret_cast<sockaddr*>(&address), sizeof(address)) < 0) {
        close(descriptor);
        throw std::runtime_error(std::strerror(errno));
    }
    socklen_t size = sizeof(address);
    getsockname(descriptor, reinterpret_cast<sockaddr*>(&address), &size);
    port = ntohs(address.sin_port);
    return descriptor;
}

std::string receive_packet(int descriptor, sockaddr_in& sender) {
    char buffer[4096]{};
    socklen_t size = sizeof(sender);
    const auto count = recvfrom(
        descriptor, buffer, sizeof(buffer), 0,
        reinterpret_cast<sockaddr*>(&sender), &size
    );
    if (count < 0) {
        throw std::runtime_error(std::strerror(errno));
    }
    return std::string(buffer, static_cast<std::size_t>(count));
}

std::string receive_type(int descriptor, std::string_view type) {
    for (int attempt = 0; attempt < 8; ++attempt) {
        sockaddr_in sender{};
        const auto packet = receive_packet(descriptor, sender);
        if (packet.find("\"type\":\"" + std::string(type) + "\"") != std::string::npos) {
            return packet;
        }
    }
    throw std::runtime_error("expected packet type was not received");
}

}  // namespace

int main() {
    using namespace gar::stream::rx;

    std::uint16_t server_port = 0;
    const int server = create_server(server_port);
    std::mutex mutex;
    std::condition_variable changed;
    std::vector<Source> latest;
    SourceBrowser browser(
        "rx-test", 0, 5600,
        {DiscoveryPeer{"127.0.0.1", server_port}},
        [&](std::vector<Source> sources) {
            {
                std::lock_guard lock(mutex);
                latest = std::move(sources);
            }
            changed.notify_all();
        }
    );
    browser.start();

    sockaddr_in browser_address{};
    const auto query = receive_packet(server, browser_address);
    assert(query.find("\"type\":\"source_query\"") != std::string::npos);
    const std::string announcement =
        "{\"protocol\":\"gar-stream/1\",\"type\":\"source_announce\","
        "\"source_id\":\"tx-one\",\"source_name\":\"Test TX\","
        "\"transport\":\"rtp/udp\",\"encoding\":\"JPEG\","
        "\"control_port\":" + std::to_string(server_port) + "}";
    assert(sendto(
        server, announcement.data(), announcement.size(), 0,
        reinterpret_cast<sockaddr*>(&browser_address), sizeof(browser_address)
    ) == static_cast<ssize_t>(announcement.size()));

    {
        std::unique_lock lock(mutex);
        assert(changed.wait_for(lock, std::chrono::seconds(2), [&] { return !latest.empty(); }));
        const Source expected{"tx-one", "Test TX", true};
        assert(latest.front() == expected);
    }
    assert(browser.select_source("tx-one"));
    const auto request = receive_type(server, "stream_request");
    assert(request.find("\"receiver_id\":\"rx-test\"") != std::string::npos);
    assert(request.find("\"stream_port\":5600") != std::string::npos);

    browser.stop();
    close(server);
    return 0;
}
